python_modules/dagster/dagster/_core/definitions/events.py

```diff
@@ -249,6 +250,116 @@ def __eq__(self, other: object) -> bool:
         )
 
 
+@whitelist_for_serdes
+class AssetFailedToMaterializeReason(Enum):
+    COMPUTE_FAILED = "COMPUTE_FAILED"  # The step to compute the asset failed
+    UPSTREAM_COMPUTE_FAILED = (
+        "UPSTREAM_COMPUTE_FAILED"  # An upstream step failed, so the step for the asset was not run
+    )
+    SKIPPED_OPTIONAL = "SKIPPED_OPTIONAL"  # The asset is optional and was not materialized
+    UPSTREAM_SKIPPED = "UPSTREAM_SKIPPED"  # An upstream asset is optional and was not materialized, so the step for the asset was not run
+    USER_TERMINATION = "USER_TERMINATION"  # A user took an action to terminate the run
+    UNEXPECTED_TERMINATION = (
+        "UNEXPECTED_TERMINATION"  # An external event resulted in the run being terminated
+    )
+    UNKNOWN = "UNKNOWN"
+
+
+# The asset can fail to materialize in two ways, an unexpected/unintentional failure that should update
+# the global state of the asset to failed, and one that indicates that the asset not materializing
+# is expected (like an optional asset, user canceled the run)
+MATERIALIZATION_ATTEMPT_FAILED_TYPES = [
+    AssetFailedToMaterializeReason.COMPUTE_FAILED,
+    AssetFailedToMaterializeReason.UPSTREAM_COMPUTE_FAILED,
+    AssetFailedToMaterializeReason.UNEXPECTED_TERMINATION,
+    AssetFailedToMaterializeReason.UNKNOWN,
+]
+
+MATERIALIZATION_ATTEMPT_SKIPPED_TYPES = [
+    AssetFailedToMaterializeReason.SKIPPED_OPTIONAL,
+    AssetFailedToMaterializeReason.UPSTREAM_SKIPPED,
+    AssetFailedToMaterializeReason.USER_TERMINATION,
+]
+
+
+@whitelist_for_serdes(
+    storage_field_names={"metadata": "metadata_entries"},
+    field_serializers={"metadata": MetadataFieldSerializer},
+)
+@record_custom
+class AssetFailedToMaterialize(EventWithMetadata, IHaveNew):
```

Comment:
the lack of a noun here that describes the event is throwing me for a bit - all the other things like this (AssetMaterialization/Observaion/CheckEvaluation/CheckResult) are nouns, this is a verb. I know why we're avoiding AssetMaterializationFailure but it still reads oddly..

Reply:
Yeah, I've been running into a similar issue in most of these PRs where there will be a function like `fetch_asset_materializations` and a similar name for the failure event doesn't quite work. I don't have any better ideas though

Reply:
The closest thing to a noun I've come up with is `AssetMaterializationMiss`, or `AssetMiss`. Because what we're really logging is a record that marks that we _know_ it's missing and info about _why_ its missing. Could also append "log" to make it more noun-y, because ultimately it's a log about the _lack_ of an event right.
`LoggedAssetMiss`. `AssetMissingLog`. `AssetMaterializationAbort`? Although maybe that last one is too perscriptive in this case.
Thoughts?

Reply:
* `AssetNonMaterialization`
* `AssetMaterializationGap`
* `AssetMaterializationAbsence`
* `AssetMissingMaterialization`
* `AssetUnfulfilledMaterialization`